source/extensions/clusters/aggregate/cluster.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include "envoy/config/cluster/aggregate/cluster.pb.h"
+#include "envoy/config/cluster/aggregate/cluster.pb.validate.h"
+
+#include "common/upstream/cluster_factory_impl.h"
+#include "common/upstream/upstream_impl.h"
+
+#include "extensions/clusters/aggregate/lb_context.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace Aggregate {
+
+class Cluster : public Upstream::BaseDynamicClusterImpl, Upstream::ClusterUpdateCallbacks {
+public:
+  Cluster(const envoy::api::v2::Cluster& cluster,
+          const envoy::config::cluster::aggregate::ClusterConfig& config,
+          Upstream::ClusterManager& cluster_manager, Runtime::Loader& runtime,
+          Runtime::RandomGenerator& random,
+          Server::Configuration::TransportSocketFactoryContext& factory_context,
+          Stats::ScopePtr&& stats_scope, bool added_via_api);
+
+  // Upstream::Cluster
+  Upstream::Cluster::InitializePhase initializePhase() const override {
+    return Upstream::Cluster::InitializePhase::Secondary;
+  }
+
+  // Upstream::ClusterUpdateCallbacks
+  void onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) override;
+  void onClusterRemoval(const std::string& cluster_name) override;
+
+  void refresh() {
+    refresh([](const std::string&) { return false; });
+  }
+
+private:
+  // Upstream::ClusterImplBase
+  void startPreInit() override;
+
+  void refresh(std::function<bool(const std::string&)>&& deleting);
+  std::pair<Upstream::PrioritySetImpl,
+            std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>>>
+  linearizePrioritySet(std::function<bool(const std::string&)> deleting);
+
+  class LoadBalancerImpl : public Upstream::LoadBalancerBase {
+  public:
+    LoadBalancerImpl(Cluster& parent,
+                     std::pair<Upstream::PrioritySetImpl,
+                               std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>>>&&
```

Comment:
might be useful to typedef some of this to make it easier to read

Reply:
Added a typedef